### PR TITLE
Adding ridge regularization in Nonnegative Parafac HALS, with some needed documentation

### PR DESCRIPTION
**Pull request by cohenjer**
This PR adds the option to impose ridge regularization in nonnegative CP decomposition with the HALS algorithm. It uses the ridge coefficient already implemented in the HALS solver. Ridge regularization is very useful because it allows to ask for sparsity on a selected number of modes using l1 regularization and applying l2 regularization elsewhere. It also allows to prune spurious components.

The PR also adds a callback interface (without removing the return error API) to be able to track the loss function (which should account for regularization including sparsity). The documentation of the function has also been improve to show the exact cost function minimized by the HALS algorithm.

This has been tested locally, I added a unit test for ridge and sparse regularized nCPD.

I also corrected a typo from an earlier PR [at line 320 of _nn_cp.py](https://github.com/tensorly/tensorly/compare/main...cohenjer:tensorly:ridge_ncp?expand=1#diff-ee80cc538fadb543db3c79fceeab03b141cc96159788b56e218eb6ac83aedf5eR330) where nn_modes should have been used instead of n_modes (sparsity is not implemented for unconstrained modes)

### CODE REVIEW

File: tensorly/decomposition/_nn_cp.py [hunk 1]
```diff
@@ -294,17 +333,36 @@ def non_negative_parafac_hals(
...
+
+    if callback is not None:
+        cp_tensor = CPTensor((weights, factors))
+        loss = (tl.norm(tensor - tl.cp_to_tensor(cp_tensor)) ** 2) / 2
+        rec_error = tl.sqrt(2 * loss) / norm_tensor
+        for mode, factor in enumerate(factors):
+            loss += ridge_coefficients[mode] ** 2 * tl.norm(
+                factor
+            ) ** 2 + sparsity_coefficients[mode] * tl.sum(tl.abs(factor))
+        callback(cp_tensor, rec_error, loss)
```
Comment (JeanKossaifi) [inline tensorly/decomposition/_nn_cp.py]: This is very specific computation just to pass to a callback no? What are the actual cases? I'd almost just have a conditional here if needed instead.

**Review by JeanKossaifi — COMMENTED** [review]
Thank you @cohenjer ! Looks great, it's super useful to have the ridge regularization option. I left some comments - mostly looking at minimizing code complexity where possible.

File: tensorly/decomposition/_nn_cp.py [hunk 2]
```diff
@@ -331,23 +389,45 @@ def non_negative_parafac_hals(
...
 
+            if callback is not None:
+                cp_tensor = CPTensor((weights, factors))
+                retVal = callback(cp_tensor, rec_error, loss)
+
+                if retVal is True:
+                    if verbose:
+                        print("Received True from callback function. Exiting.")
+                    break
+        if tol:
```
[the remark this answers is not in the document]
Comment (cohenjer) [reply tensorly/decomposition/_nn_cp.py:416]: Hi Jean, 

To answer both your comments, indeed I was a bit conflicted here, because the previous code only returned the reconstruction error, but the algorithm is designed to minimize the objective function (including regularization). I did not want to break previous behavior and therefore still return the reconstruction error, but also return the loss (objective function) which I find also quite useful.

I guess we can also just return the factors and mttkrp and not the reconstruction error nor the loss, and let the user compute them manually if he wants to return the loss/rec error. This way in case a callback is used, the runtime would not necessarily be slower.  The code would also be much simpler. We also already have error calc function knowing mttkrp somewhere in tensorly, so that would not be hard to use.

I could add the code for callback loss function in the documentation/examples. 

To summarize:
- Either we simplify the callback by just passing factors and mttkrp. It makes it slightly harder for the user to return rec error or objective function, but we can provide an example.
- Or we pass the error and/or objective function as a callback argument, as I did, in which case we could, as you suggest, return only the rec_error. The code would be very similary to the previous version in that case.

[the remark this answers is not in the document]
Comment (cohenjer) [reply tensorly/decomposition/_nn_cp.py:416]: > Another comment is that this conflates the objective and the reconstruction error - to me the reconstruction error should not include the ridge regularization, unlike the objective function.

This is actually exactly what I did I think!

**Comment by cohenjer** [comment]
@JeanKossaifi I followed your recommendation and simplified all the callback calls to simply use the unnormalized reconstruction error in the signature, along with the CP tensor. This way the user can easily compute the loss function, and I believe it is consistent with the CP decomposition loss.
